Issue 104 PR for Listing Whole view in the Directory View

k10blogger opened the pull request:
Here is an attempt that takes a one time load time penalty and loads the entire repository in the directory.
The plan is that once the repository is loaded the javascript shall take care of collapsing of the folders. 

Right now behaviour is that on click the folder is zoomed in.
All comparison and everything else work as expected.

Following is a snapshot of how it looks:
![image](https://user-images.githubusercontent.com/4973182/92032636-1b71cf00-ed88-11ea-8f85-f12fa76d6293.png)

michael-o commented:
This should not be the default. By default, only one level should be reachable. A config option would be required her.

ams-tschoening commented:
> Right now behaviour is that on click the folder is zoomed in.

Just to be sure, because your screenshot looks different: The default is to load all contents, but show all directories as collapsed? So users need to click each directory to expand it and that happens instantly because everything has been loaded already?

I agree with @michael-o that this behaviour should be configured, but otherwise the approach looks fine in general to me. Do you have some before/after times for your concrete repos how long initial loading takes?

ams-tschoening commented on include/svnlook.php:
```diff
@@ -1039,11 +1039,11 @@ function getList($path, $rev = 0, $peg = '') {
 
 		// Sort the entries into alphabetical order
-		usort($curList->entries, '_listSort');
+		/*usort($curList->entries, '_listSort');*/
```
Why did you remove the sorting? Many user interfaces sort directory and file entries according some criteria instead of relying on arbitrary order of the underlying file system or whatever.

k10blogger replied:
When i take the xml output out it has the correct order of directories and the files. First the directory then the files under it and then so on. 
When i do this sort it puts all the directories at top and then fhe files in an unrecognizable order. Where the xml has the proper order printed out. So i removed the sorting.

k10blogger replied:
I did some investigation. This function will not work in this case as it is fed the whole list of directories and files by this method. 
Previously it was being fed level by level. First the root level, then the folder that is clicked level and so on... So it was never organizing everything, only each level. 
Alphabetically sorting the entire list while still keeping level information will be tough as the xml output doesnt provide any level information. We calculate level by slicing the folders for each file that is fed.

ams-tschoening replied:
So whats your conclusion, remove the entire thing as I suggested? I'm fine with applying alphabetical sort always by default, without any further config. You only need to confirm that things are sorted alphabetically in your case with loading everything recursively AND with the former behaviour loading one level only.

k10blogger replied:
I am not sure in the full recursive scenario i will be able to arrange alphabetically. Let me try it first i will come back here after that. Then we can come to conclusion.

ams-tschoening replied:
We might be talking about different things: I DO NOT recommend alphabetical sorting ACROSS dirs and files. It should be dirs first, than files and within both alphabetically always. That's how I understood your following sentence:

> First the directory then the files under it and then so on.

But the XML doesn't care about alphabetical sort within each group?

k10blogger replied:
Yes the XML doesn't care about sorting. There is no group in xml. Sorting there does sorting across dirs and files. There is no way to avoid that. 
Unless we do further processing on it. There is no way for recursive xml to first sort the dirs and then the files like it is being currently done. 

Here is a snapshot of xml. 
![image](https://user-images.githubusercontent.com/4973182/92223949-d308fd00-eebe-11ea-8b5f-9cd4714442d0.png)

k10blogger replied:
@ams-tschoening Now i understood what you meant. I have made the sorting by alphabetical order by default for when the repository load all at once.

k10blogger commented:
> Just to be sure, because your screenshot looks different: The default is to load all contents, but show all directories as collapsed? So users need to click each directory to expand it and that happens instantly because everything has been loaded already?

Yes that is the intended approach. 

Okay will take some time but will put it as a configurable option. 
I will also do the following:

1. Before and after time for repo loading
2. Yes restriction are there. But i didnt check. Will do that as well.

@michael-o but are you okay with this approach?

michael-o commented:
Will go through tomorrow.

k10blogger commented:
In original:
I took a time measurement. It varies from path to path. Like for the first path.
7.467 seconds. Then it stays between 9 - 12 seconds for other paths.
![image](https://user-images.githubusercontent.com/4973182/92201532-7562b980-ee9a-11ea-9cc9-650d97d808aa.png)

In full directory loading same repository takes 50 seconds flat:
![image](https://user-images.githubusercontent.com/4973182/92201791-0c2f7600-ee9b-11ea-94ce-882800bc2a29.png)

So 5 clicks on the original equals one on the full directory loading.

The load time calculation starts in the listing.php once it goes in the `if($rep)` section.

ams-tschoening commented:
Those times look pretty much like what I see for some of my repos. The important difference is that without loading everything at once, users get to see something earlier, which might be preferred and should justify an additional config to switch between both behaviours.

Why things are that slow has been discussed in #78 already. When I look at ProcMon with my repo, it's again pretty clear that most of the time accessing shell tools `svn.exe` and `svnauthz.exe` is the problem.

[ProcMon WebSVN.zip](https://github.com/websvnphp/websvn/files/5173239/ProcMon.WebSVN.zip)

k10blogger commented:
Added Config for Load Repository. Parameter to set to TRUE all repo is loaded or original behaviour.
Same check added in the html template.
Took some time and study but fixed the Javascript. (Clicking on folder icon collapses or opens the paths)
Removed load times variables that were added for testing. Though we should think of adding it as additional information?

ams-tschoening reviewed (changes requested):
Thinks look good in general, but I suggested lots of improvements to the code style. Makes changes easier to read and maintain. So please read all those changes and renames first and AFTERWARDS have a look at a different implementation for `collapseAllDir` using CSS-selectors and jQuery. With this order you get a better idea of why I suggested the newlines etc., I would like to see the difference yourself first.

Regarding your newly added JS, I suggest a different approach with not calling a function to do things, but using a jQuery handler again, like you did the last time already as well. This saves you from iterating individual rows and use CSS-selectors instead, that's what they are for in the end.

k10blogger commented on include/configclass.php:
```diff
@@ -672,6 +672,8 @@ class WebSvnConfig {
 	var $authz = null;
 	var $blockRobots = false;
 
+	var $loadallrepo = false;
```
Done. changed $loadallrepo to $loadAllRepos everywhere.

k10blogger commented on include/configclass.php:
```diff
@@ -1555,6 +1557,14 @@ function getOpenTree() {
 		return $this->openTree;
 	}
 
+	function setLoadAllRepository($flag) {
```
Done in both  set and show functions.

k10blogger commented on javascript/collapsible.js:
```diff
@@ -18,10 +18,69 @@ function collapseAllGroups() {
   }
 }
+function collapseAllDir() {
+    for (var i = 1; i < tableRows.length; i++) {
```
This is something I didn't know.  Used `let` wherever possible.

ams-tschoening replied:
That's some of the basics I recently learned as well only and wanted to share. Makes code more robust in the end.

k10blogger commented on javascript/collapsible.js:
```diff
@@ -18,10 +18,69 @@ function collapseAllGroups() {
+    for (var i = 1; i < tableRows.length; i++) {
+        var strclass = tableRows[i].title;
+        var res = strclass.split(" ");
+        for (var j = i+1; j<tableRows.length; j++) {
```
For each block like if|fow|while added newline before. I have covered in most of the places now.

k10blogger commented on javascript/collapsible.js:
```diff
@@ -18,10 +18,69 @@ function collapseAllGroups() {
 });
 
+$("tr").find("td.path").on("click",function(event) {
+    event.stopPropagation();
```
Done new lines before and after `event` and code blocks.

k10blogger commented on javascript/collapsible.js:
```diff
@@ -18,10 +18,69 @@ function collapseAllGroups() {
+    if ($target.closest("tr").next().attr("title") != undefined) {
+        var strclasscheck = $target.closest("tr").next().attr("title");
+        var rescheck = strclasscheck.split(" ");
+        var performaction = $target.closest("tr").attr("customaction") == 'close'? 'open' : 'close';
```
Changed to `let` and new lines after blocks.

k10blogger commented on javascript/collapsible.js:
```diff
@@ -18,10 +18,69 @@ function collapseAllGroups() {
+    var strclass = $target.closest("tr").attr("title");
+    var res = strclass.split(" ");
+    // On Click Check
+    if ($target.closest("tr").next().attr("title") != undefined) {
```
This is a really good advice exit early. I have done this. Will keep this in mind for all future commits.

k10blogger commented on listing.php:
```diff
@@ -191,22 +191,166 @@ function showDirFiles($svnrep, $subs, $level, $limit, $rev, $peg, $listing, $ind
 
-function showTreeDir($svnrep, $path, $rev, $peg, $listing) {
-	global $vars, $config;
+function showAllDirFiles($svnrep, $path, $rev, $peg, $listing, $index, $treeview = true) {
```
Changed the treeview to treeView. Yeah i will keep this in mind as well. As i took this piece of code from existing code. But you are right need to improve when writing something new otherwise code style will never improve.

k10blogger commented on listing.php:
```diff
@@ -191,22 +191,166 @@ function showDirFiles($svnrep, $subs, $level, $limit, $rev, $peg, $listing, $ind
+		$parentPath = $subs;
+		unset($parentPath[count($parentPath) - 2]);
+		$parentPath = implode('/', $parentPath);
+		if ($rep->hasReadAccess($parentPath, false)) {
```
Done. Added new line before the if block.

k10blogger commented on listing.php:
```diff
@@ -191,22 +191,166 @@ function showDirFiles($svnrep, $subs, $level, $limit, $rev, $peg, $listing, $ind
+	}
+	$openDir = false;
+	$logList = $svnrep->getList($path, $rev, $peg);
+	if ($logList) {
```
Done. Changed at two locations one at `$loglist` and other at `accesss`.

k10blogger commented on listing.php:
```diff
@@ -191,22 +191,166 @@ function showDirFiles($svnrep, $subs, $level, $limit, $rev, $peg, $listing, $ind
+			$access = ($isDir)	? $rep->hasReadAccess($path.$file, false)
+								: $accessToThisDir;
+
+			if ($access) {
```
Done. Works the same. (The newline came in with the previous section of the code cant take credit of adding it :) ).

k10blogger commented on listing.php:
```diff
@@ -191,22 +191,166 @@ function showDirFiles($svnrep, $subs, $level, $limit, $rev, $peg, $listing, $ind
+					$listvar['filetype'] = strtolower(strrchr($file, '.'));
+					$openDir = false;
+				}
+				$listvar['isDir'] = $isDir;
```
Done. new line before this code block and after the code block as well.

k10blogger commented on listing.php:
```diff
@@ -191,22 +191,166 @@ function showDirFiles($svnrep, $subs, $level, $limit, $rev, $peg, $listing, $ind
+				$listvar['openDir'] = $openDir;
+				$listvar['path'] = $path.$file;
+				$tempelements = explode('/',$file);
+				if ($tempelements[count($tempelements)-1] === "")
```
Done.

k10blogger commented on listing.php:
```diff
@@ -191,22 +191,166 @@ function showDirFiles($svnrep, $subs, $level, $limit, $rev, $peg, $listing, $ind
+					$listvar['level'] = ($treeview) ? $lastindexfile : 0;
+					$listvar['filename'] = $tempelements[$lastindexfile];
+					$listvar['classname'] = '';
+					for($n=0; $n<$lastindexfile; $n++)
```
Done. Before and after for block.

k10blogger commented on listing.php:
```diff
@@ -191,22 +191,166 @@ function showDirFiles($svnrep, $subs, $level, $limit, $rev, $peg, $listing, $ind
+						$listvar['last_i_node'][$n] = false;
+						$listvar['classname'] = $listvar['classname'].$tempelements[$n].' ';
+					}
+					$listvar['classname'] = $listvar['classname'].$tempelements[$lastindexfile];
```
Done before the code block. Didnt add after the block as the block ends.

k10blogger commented on listing.php:
```diff
@@ -191,22 +191,166 @@ function showDirFiles($svnrep, $subs, $level, $limit, $rev, $peg, $listing, $ind
+					$listvar['level'] = ($treeview) ? $lastindexfile : 0;
+					$listvar['filename'] = $tempelements[$lastindexfile];
+					$listvar['classname'] = '';
+					for($n=0; $n<$lastindexfile; $n++)
```
Done before and after the `for` code block.

k10blogger commented on listing.php:
```diff
@@ -191,22 +191,166 @@ function showDirFiles($svnrep, $subs, $level, $limit, $rev, $peg, $listing, $ind
+					}
+					$listvar['last_i_node'][$lastindexfile] = true;
+				}
+				if ($isDir) {
```
Done before and after the `if-else` block.

k10blogger commented on listing.php:
```diff
@@ -191,22 +191,166 @@ function showDirFiles($svnrep, $subs, $level, $limit, $rev, $peg, $listing, $ind
+					$listvar['fileurl'] = urlForPath($path.$file, createDifferentRevAndPegString($passrev, $peg));
+				}
+				$listvar['filelink'] = '<a href="'.$listvar['fileurl'].'">'.$listvar['filename'].'</a>';
+				if ($isDir) {
```
Done. Before and after the `if-else` block.

k10blogger commented on listing.php:
```diff
@@ -191,22 +191,166 @@ function showDirFiles($svnrep, $subs, $level, $limit, $rev, $peg, $listing, $ind
-	// For file, the last element in the subs is the file name.
-	// Therefore, it is always count($subs) - 2
-	$limit = count($subs) - 2;
+				if ($treeview) {
```
Changed treeview to treeView and newline added.

k10blogger commented on listing.php:
```diff
@@ -191,22 +191,166 @@ function showDirFiles($svnrep, $subs, $level, $limit, $rev, $peg, $listing, $ind
+				if ($treeview) {
+					$listvar['compare_box'] = '<input type="checkbox" name="compare[]" value="'.escape($path.$file).'@'.$passrev.'" onclick="enforceOnlyTwoChecked(this)" />';
+				}
+				if ($config->showLastModInListing()) {
```
Done before and after `if`.

k10blogger commented on listing.php:
```diff
@@ -191,22 +191,166 @@ function showDirFiles($svnrep, $subs, $level, $limit, $rev, $peg, $listing, $ind
+					$listvar['date'] = $entry->date;
+					$listvar['revurl'] = $config->getURL($rep, $path.$file, 'revision').$isDirString.createRevAndPegString($entry->rev, $peg ? $peg : $rev);
+				}
+				if ($rep->isDownloadAllowed($path.$file)) {
```
Done before and after `if`.

k10blogger commented on listing.php:
```diff
@@ -191,22 +191,166 @@ function showDirFiles($svnrep, $subs, $level, $limit, $rev, $peg, $listing, $ind
+					$listvar['downloadplainurl'] = '';
+					$listvar['downloadurl'] = '';
+				}
+				if ($rep->isRssEnabled()) {
```
Done before and after `if`.

k10blogger commented on listing.php:
```diff
@@ -191,22 +191,166 @@ function showDirFiles($svnrep, $subs, $level, $limit, $rev, $peg, $listing, $ind
+
+function showTreeDir($svnrep, $path, $rev, $peg, $listing) {
+	global $vars, $config;
+	if ($config->showLoadAllRepository()) {
```
Done before `if`.

k10blogger commented on templates/BlueGrey/directory.tmpl:
```diff
@@ -124,4 +128,13 @@ e-node=<img class="icon" border="0" width="24" height="22" src="[websvn:locwebsv
+  <script type="text/javascript" src="[websvn:locwebsvnhttp]/javascript/collapsible.js"></script>
+      <script type="text/javascript">
+      //<![CDATA[
+      collapseAllDir();
```
Used the same `each` concept in a slightly different way. Removed the `collapseAllDir` implementation completely.

k10blogger commented on templates/Elegant/directory.tmpl:
```diff
@@ -134,4 +138,13 @@ e-node=<img src="[websvn:locwebsvnhttp]/templates/Elegant/images/blank.png" alt=
+    [websvn-test:loadalldir]
+    <script type="text/javascript" src="[websvn:locwebsvnhttp]/javascript/jquery.min.1.9.1.js"></script>
+    <script type="text/javascript" src="[websvn:locwebsvnhttp]/javascript/collapsible.js"></script>
+        <script type="text/javascript">
```
Removed for jQuery-handler from all 3 templates.

k10blogger commented on templates/calm/directory.tmpl:
```diff
@@ -138,4 +142,13 @@ e-node=<img src="[websvn:locwebsvnhttp]/templates/calm/images/e-node.png" alt="[
+[websvn-test:loadalldir]
+<script type="text/javascript" src="[websvn:locwebsvnhttp]/javascript/jquery.min.1.9.1.js"></script>
+<script type="text/javascript" src="[websvn:locwebsvnhttp]/javascript/collapsible.js"></script>
+    <script type="text/javascript">
```
Removed for jQuery-handler from all 3 templates.

k10blogger commented on listing.php:
```diff
@@ -191,22 +191,166 @@ function showDirFiles($svnrep, $subs, $level, $limit, $rev, $peg, $listing, $ind
+		$vars['compare_box'] = ''; // Set blank once in case tree view is not enabled.
+		return showAllDirFiles($svnrep, $path, $rev, $peg, $listing, 0, $config->treeView);
 	}
+	else {
```
Done this in $access and $loglist.

k10blogger commented on javascript/collapsible.js:
```diff
@@ -18,10 +18,69 @@ function collapseAllGroups() {
       tableRows[i].style.display = 'none';
   }
 }
+function collapseAllDir() {
```
Removed this function completely in favor of jquery handler that loops each row.

k10blogger commented on include/distconfig.php:
```diff
@@ -166,6 +166,14 @@
+// This options loads the entire directory in one go and allows to browse without delay.
+// By default all will be collapsed to root directory and can be expanded.
+// This superseeds the setAlphabeticOrder and everything will be shown in alphabetic 
+// order by default.
```
This is an example how it looks. It is alphabetically sorted by default as shown.
![image](https://user-images.githubusercontent.com/4973182/92430756-507c8800-f1b3-11ea-8b75-4d0fce123850.png)

When directory is expanded then the files and folders under it are also alphabetically arranged by default.
![image](https://user-images.githubusercontent.com/4973182/92430804-7a35af00-f1b3-11ea-99e6-be79b3b60262.png)

Yes removed the sorting call from when loadAllRepos is shown. But it retains the behavior when it is false.

ams-tschoening replied:
OK, so when you talk about alphabetic sort order, you always mean that grouping of dirs and files is NOT supported anymore. In that case please add an additional sentence to tell so in the docs of your config. This can be improved later in the worst case and is known until then.

> // The alphabatecal order is applied to all directory and files. 

vs.

> // The alphabetical order is applied to all directory and files. This means that grouping of all dirs together and all files together is NOT supported currently!

k10blogger replied:
Wrote something like the following. 
```
// By default, WebSVN loads parent path directories and then on user click other,
// This options loads the entire directory in one go and allows to browse without delay.
// By default all will be collapsed to root directory and can be expanded.
// The performance will be impacted as it takes time to load up all the things in the 
// repository. Once loaded directory exapansion is instantaneous.
// The alphabetical order is applied to all directory and files. 
// This means that grouping of all dirs together and all files together is NOT supported currently!
// The files and directories are shown as is with a mixture of files and folders.
```

k10blogger commented on listing.php:
```diff
@@ -191,22 +191,166 @@ function showDirFiles($svnrep, $subs, $level, $limit, $rev, $peg, $listing, $ind
+			$index++;
+		}
+	}
+	$openDir = false;
```
Added newline before and after this code block.

ams-tschoening commented on javascript/collapsible.js:
```diff
@@ -19,9 +19,88 @@ function collapseAllGroups() {
 
+var strclass = '';
+var res = '';
+$("tbody > tr").each(function() {
```
What are you trying to achieve here?

Working with global variables like in this case, accessed by different handlers is a bad implementation. While I suggested using jQuery-handlers, that doesn't mean you need to remove functions entirely. Keep the code of the handlers in functions as much as possible and call those in the handlers. This is more readable than what you have now, functions can be documented easier etc.

So please describe what you are trying to achieve with this and the other handler and I'll try to find some better implementation without those global variables and twi functions being called by the handler.

ams-tschoening replied:
Regarding code style again: It doesn't make much sense to use code like the following:

	$("tbody > tr").each(function() {

		some stuff
	});

If you want the improved readability by the newline, simply use a style like the following:

	$("tbody > tr").each(function()
	{
		some stuff
	});

Your former style is more used by people who do NOT want the extra newline and save some visual space in the end. So the latter is more standard and widely known.

k10blogger replied:
There are two handlers 1 to run at the time of page load and 1 to run at the time on user operation.
1 - Page Load `$("tbody > tr").each(function() { });`
When the entire page is loaded with the data this handler will go through each of the table row. It will discard the rows that dont have `title` marked. The title attribute is used to generate the path and level information from the the file name that has been provided. From log file we get data as:
```
branches/ -> title = "branches"
branches/folder1/item1.txt - > title = "branches folder1"
branches/folder1.item2.txt - > title = "branches folder1"
branches/item3.txt - > title = "branches"
```
Aim of this handler to first identify unique folders at the root level and collapse all the items that are marked with that folder name. Here in this case branches. 
This will keep only root folders visible rest all will be collapsed.
This is how we get the first page with only root folders.

2 - User Operation - On Click Operation
This is used to expand the folders that are collapsed during the page load.
It is to attach to `td.path` as it is the dominant element visually. 
As on page load everything is closed the customaction is close. This is the attribute is used to track the operation performed on the folder.
When we click on say branches. 
It traverse through every row below it and keep making rows visible which all have the `branches` in the title. Meaning all files, and folders under branches. 
It exapands the subfolder in the branches as well. Giving an entire tree view. It marks all items with customaction as open as that is the action that has been performed.
Same logic is used to collapse or expand the subfolders udner the root folders. 

For the first handler as the handler is called each row there is a need to store value for a scope larger than the function call to track what is collapsing. Hence the variable
```
var strclass = '';
var res = ''; 
```
I am not able to figure it out how to do it some other way. Would be good know if a better way exists.

ams-tschoening replied:
While I didn't understand all of this, I have the feeling you are re-implementing CSS-selectors in the end and should instead simply use those in combination with registering click-events on individual dirs. There are selectors for e.g. searching individual words in HTML-attributes already, which seems to be exactly what you are doing with `title`:

	/* <a> elements whose class attribute contains the word "logo" */
	a[class~="logo"] {
	  padding: 2px;
	}

https://developer.mozilla.org/en-US/docs/Web/CSS/Attribute_selectors

State to distinguish what to do shouldn't be encoded in global variables unless absolutely necessary, which it doesn't seem to be in your case. When you know if to show/hide things, you can as well register click events on those elements to do something. This way you don't need state in variables, the pure fact that a click-handler is available doing what you want with this element is enough.

I'll try to have a closer look at the weekend.

k10blogger replied:
I looked at the link that you provided and after some search. I implemented an even simpler change. But that meant that i had to unified the `directory.tmpl` across all the three templates. Have a look at the new commit.

ams-tschoening commented on javascript/collapsible.js:
```diff
@@ -19,9 +19,88 @@ function collapseAllGroups() {
+
+    let strclasscheck = $target.closest("tr").next().attr("title");
+    let rescheck = strclasscheck.split(" ");
+    let performaction = $target.closest("tr").attr("customaction") == 'close'? 'open' : 'close';
```
All of the above names read badly again and should be camel-cased:

    let strClassCheck =
    let resCheck =
    let performAction =

k10blogger replied:
Changed the variable name to camel-cased. For the handlers.
Also spaced out the entire Js file to try to make it consistent.

ams-tschoening commented on javascript/collapsible.js:
```diff
@@ -1,90 +1,107 @@
-      }
+function toggleGroup(groupName) 
+{
+
```
Remove those newlines, things are pretty readable already like you have them now.

	function toggleGroup(groupName) 
	{

		some stuff

	}

	vs.

	function toggleGroup(groupName) 
	{
		some stuff
	}

The former is only a waste of visual space in many cases. To get that visual space, you have moved the `{` down the line already.

k10blogger replied:
I have done this change in the new commit.

ams-tschoening commented on javascript/collapsible.js:
```diff
@@ -1,90 +1,107 @@
+
+    for (var i = 0; i < tableRows.length; i++) 
+    {
+        if (tableRows[i].title != '')
```
You should avoid avoid statements like `if` without brackets as much as possible. It's too easily to wrongly change them:

	if (tableRows[i].title != '')
		tableRows[i].style.display = 'none';

	vs.

	if (tableRows[i].title != '')
		tableRows[i].style.display = 'none';
		tableRows[i].style.width = '40px';

	if (tableRows[i].title != '')
	{
		tableRows[i].style.display = 'none';
	}

	vs.

	if (tableRows[i].title != '')
	{
		tableRows[i].style.display = 'none';
		tableRows[i].style.width = '40px';
	}

Only the last two examples work always as intended.

k10blogger replied:
I have changed the if and else statement. I will keep this in mind if i come across more such thing in other section of the code. It is better to clean them then to keep it as is.

ams-tschoening commented:
I had a look at this PR and it didn't work for me for various reasons. So I re-implemented the JS to what I think is better to maintain and more clear how it works. In the end, all we want is toggling rows in the table and that seems to be possible without global variables, custom attributes etc. The only problem is finding child elements of parents and such, but that is possible using selectors and some additional regular expressions.

@k10blogger Please have a look at #122 and the associated branch and test things from there. I'm closing here for now, because your work is already contained in the other PR.
